cpp/src/bot.cpp

```diff
@@ -0,0 +1,127 @@
+#include "./bot.hpp"
+#include <algorithm>
+
+#ifdef RANGES_SUPPORT
+#undef RANGES_SUPPORT
+#endif
+
+#if _HAS_CXX20
+#define RANGES_SUPPORT 1
+#include <ranges>
+#endif
+
+// Internal
+std::ostream &operator<<(std::ostream &os, Card const &card)
+{
+    return os << Card::ToStr(card) << std::endl;
+}
+
+std::ostream &operator<<(std::ostream &os, PlayPayload const &payload)
+{
+    os << "Player ID : " << payload.player_id << std::endl;
+    os << "Player IDs : ";
+    for (auto const &player : payload.player_ids)
+        os << player << "    ";
+
+    os << "\nCards \n";
+    for (auto card : payload.cards)
+        os << card << "   ";
+
+    os << "\nPlayed : \n";
+    for (auto card : payload.played)
+        os << card << "  \n";
+
+    os << "\nBid History : \n";
+    for (auto const &entry : payload.bid_history)
+    {
+        os << "player    : " << entry.player_id << "\n"
+           << "Bid value : " << entry.bid_value << std::endl;
+    }
+
+    // Hand history
+    for (auto const &x : payload.hand_history)
+    {
+        os << "Initiator : " << x.initiator;
+        os << "\nWinner   : " << x.winner;
+        os << "\nCards Played : ";
+        for (auto card : x.card)
+            os << card << "   ";
+        os << std::endl;
+    }
+    return os;
+}
+
+static GameState game_instance;
+
+GameState& GetGameInstance()
+{
+    return game_instance;
+}
+
+void InitGameInstance()
+{
+    // Initialization code of game state goes here, if any
+}
+
+//
+// Actual gameplay goes here
+//
+// These three functions are called in response to the http request calls /chooseTrump, /bid and /play respectively. 
+// Logic of your code should go inside respective functions
+//
+
+Suit GameState::ChooseTrump(PlayerID myid, std::vector<PlayerID> player_ids, std::vector<Card> mycards,
+                            int32_t time_remaining, std::vector<BidEntry> bid_history)
+{
+    return Suit(mycards[3].suit);
+}
+
+int GameState::Bid(PlayerID myid, std::vector<PlayerID> player_ids, std::vector<Card> mycards, int32_t time_remaining,
+                   std::vector<BidEntry> bid_history, BidState const &bid_state)
+{
+    // Either bid or pass
+    if (bid_state.defender.bid_value == -1 && bid_state.challenger.bid_value == -1)
+        return 16;
+    return 0;
```

Comment:
@TilakChad Please refrain from using  `-1` as an indicator to pass the bid. 
Check whether `bidHistory` is empty or not to pass the minimum bid.

`-1` may change.

Reply:
I've  modified that to check for empty bid history now. I was following the API docs closely, so thought that would be  the way. Other things like the order of json element weren't were exactly like as the API docs enlisted. So had to consider that too.

Reply:
@TilakChad Other things like the order of json element weren't were exactly like as the API docs enlisted. So had to consider that too.

Could you elaborate a bit more on this?

Reply:
> @TilakChad Other things like the order of json element weren't were exactly like as the API docs enlisted. So had to consider that too.
> 
> Could you elaborate a bit more on this?

I was referring to this part :
![image](https://user-images.githubusercontent.com/49703944/203919642-1f3b11d4-4deb-4c4b-9d59-822914018ede.png)

and the actual payload  : 
![image](https://user-images.githubusercontent.com/49703944/203920110-93626d04-4173-4733-8fc1-3211cedc9c8a.png)

As far as it concerns API documentation, this is an oversight. 

Faced onto this while parsing json sequentially on that order, which would make parsing a bit faster. Usual way is indexing onto the property id : value, which isn't terrible but okish.